streams/src/test/java/org/apache/kafka/streams/integration/RepartitionOptimizingIntegrationTest.java

```diff
@@ -0,0 +1,258 @@
+    private static final String AGGREGATION_TOPIC = "outputTopic_1";
+    private static final String REDUCE_TOPIC = "outputTopic_2";
+    private static final String JOINED_TOPIC = "joinedOutputTopic";
+
+    private static final int ONE_REPARTITION_TOPIC = 1;
+    private static final int FOUR_REPARTITION_TOPICS = 4;
+
+    private final Pattern repartitionTopicPattern = Pattern.compile("Sink: .*-repartition");
+
+    private Properties streamsConfiguration;
+
+
+    @ClassRule
+    public static final EmbeddedKafkaCluster CLUSTER = new EmbeddedKafkaCluster(NUM_BROKERS);
+    private final MockTime mockTime = CLUSTER.time;
+
+    @Before
+    public void setUp() throws Exception {
+        final Properties props = new Properties();
+        props.put(StreamsConfig.CACHE_MAX_BYTES_BUFFERING_CONFIG, 1024 * 10);
+        props.put(StreamsConfig.COMMIT_INTERVAL_MS_CONFIG, 5000);
+
+        streamsConfiguration = StreamsTestUtils.getStreamsConfig(
+            "maybe-optimized-test-app",
+            CLUSTER.bootstrapServers(),
+            Serdes.String().getClass().getName(),
+            Serdes.String().getClass().getName(),
+            props);
+
+        CLUSTER.createTopics(INPUT_TOPIC,
+                             COUNT_TOPIC,
+                             AGGREGATION_TOPIC,
+                             REDUCE_TOPIC,
+                             JOINED_TOPIC);
+
+
+        IntegrationTestUtils.purgeLocalStreamsState(streamsConfiguration);
+    }
+
+    @After
+    public void tearDown() throws Exception {
+        CLUSTER.deleteAllTopicsAndWait(30_000L);
+    }
+
+    @Test
+    public void shouldSendCorrectRecords_OPTIMIZED() throws Exception {
+        runIntegrationTest(StreamsConfig.OPTIMIZE,
+                           ONE_REPARTITION_TOPIC);
+    }
+
+    @Test
+    public void shouldSendCorrectResults_NO_OPTIMIZATION() throws Exception {
+        runIntegrationTest(StreamsConfig.NO_OPTIMIZATION,
+                           FOUR_REPARTITION_TOPICS);
+    }
+
+
+    private void runIntegrationTest(final String optimizationConfig,
+                                    final int expectedNumberRepartitionTopics) throws Exception {
+
+        final Initializer<Integer> initializer = () -> 0;
+        final Aggregator<String, String, Integer> aggregator = (k, v, agg) -> agg + v.length();
+
+        final Reducer<String> reducer = (v1, v2) -> v1 + ":" + v2;
+
+        final List<String> processorValueCollector = new ArrayList<>();
+
+        final StreamsBuilder builder = new StreamsBuilder();
+
+        final KStream<String, String> sourceStream = builder.stream(INPUT_TOPIC, Consumed.with(Serdes.String(), Serdes.String()));
+
+        final KStream<String, String> mappedStream = sourceStream.map((k, v) -> KeyValue.pair(k.toUpperCase(Locale.getDefault()), v));
+
+
+        mappedStream.filter((k, v) -> k.equals("B")).mapValues(v -> v.toUpperCase(Locale.getDefault())).process(() -> new SimpleProcessor(processorValueCollector));
```

Comment:
What is this branch used for?

Reply:
For testing that the optimization only moves nodes that strictly require movement for optimization.  

In this case, this branch of the `mappedStream` don't make any calls demanding a repartition operation, thus should remain in their original position in the graph.  I included this branch so I can render the topology in the visual tool to confirm this child node, in fact, does not get moved.